rust-runtime/smithy-eventstream/src/frame.rs

```diff
@@ -581,10 +556,141 @@ mod tests {
         let expected = include_bytes!("../test_data/valid_with_all_headers_and_payload").to_vec();
         assert_eq!(expected, actual);
 
-        let result = Message::read_from(&mut Bytes::from(actual))
-            .unwrap()
-            .unwrap();
+        let result = Message::read_from(&mut Bytes::from(actual)).unwrap();
         assert_eq!(message.headers(), result.headers());
         assert_eq!(message.payload().as_ref(), result.payload.as_ref());
     }
 }
+
+/// Return value from [`MessageFrameDecoder`].
+#[derive(Debug)]
+pub enum DecodedFrame {
+    /// There wasn't enough data in the buffer to decode a full message.
+    Incomplete,
+    /// There was enough data in the buffer to decode.
+    Complete(Message),
+}
+
+/// Streaming decoder for decoding a [`Message`] from a stream.
+#[non_exhaustive]
+#[derive(Default)]
+pub struct MessageFrameDecoder {
+    prelude: [u8; PRELUDE_LENGTH_BYTES_USIZE],
+    prelude_read: bool,
+}
+
+impl MessageFrameDecoder {
+    /// Returns a new `MessageFrameDecoder`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Determines if the `buffer` has enough data in it to read a full frame.
+    /// Returns `Ok(None)` if there's not enough data, or `Some(remaining)` where
+    /// `remaining` is the number of bytes after the prelude that belong to the
+    /// message that's in the buffer.
+    fn remaining_ready<B: Buf>(&self, buffer: &B) -> Result<Option<usize>, Error> {
```

Comment:
found this function name a little hard to grok. maybe `try_next_end_of_frame`? don't really love that either

Reply:
Yeah, this one was really hard to come up with a good name for. I'll go for `remaining_bytes_if_frame_available`. It's slightly better at least:
```rust
if let Some(remaining_len) = self.remaining_bytes_if_frame_available(&buffer)? {
```